libbeat/mime/byte.go

```diff
@@ -0,0 +1,71 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mime
+
+import (
+	"encoding/json"
+	"encoding/xml"
+	"net/http"
+	"strings"
+
+	"github.com/h2non/filetype"
+)
+
+const (
+	// size for mime detection, office file
+	// detection requires ~8kb to detect properly
+	headerSize = 8192
+)
+
+// DetectBytes tries to detect a mime-type based off
+// of a chunk of bytes passed into the function
+func DetectBytes(data []byte) string {
+	header := data
+	if len(data) > headerSize {
+		header = data[:headerSize]
+	}
+	kind, err := filetype.Match(header)
+	if err == nil && kind != filetype.Unknown {
+		// we have a known filetype, return
+		return kind.MIME.Value
+	}
+	// if the above fails, try and sniff with http sniffing
+	netType := http.DetectContentType(header)
+	// try and parse any sort of text as json or xml
+	if strings.HasPrefix(netType, "text/plain") {
+		if detected := detectEncodedText(data); detected != "" {
+			return detected
+		}
+	}
+	if netType == "application/octet-stream" {
+		return ""
+	}
+	return netType
+}
+
+func detectEncodedText(data []byte) string {
+	// figure out how to optimize this so we don't have to try and parse the whole payload
+	// every time
+	if json.Valid(data) {
+		return "application/json"
+	}
+	if xml.Unmarshal(data, new(interface{})) == nil {
```

Comment:
This is going to be really expensive, both in time & memory allocations.  Maybe we could replace it with an XML Decoder that checks that we have matching starting & ending elements?

Reply:
I was initially contemplating looking for `<?xml` starting blocks or `xmlns` attributes, but a bunch of xml api responses don't necessarily have either. The performance concerns are definitely on my mind too and why I threw the comment above and why this is the last of all detection conditionals. But, since this code path would only trigger when:

1. You have a filled in field you're trying to detect and
2. All other content-type analysis has failed

I figured it was fine to keep it in for now. I'm hoping to get this PR merged soonish so it doesn't block adding the processor to other beats for the ECS upgrade. The matching idea seems reasonable to me and I could do some other research to figure out alternatives, but I'm wondering if either:

1. We could keep this as is for now due to the fact that it's not necessarily going to be hit often.
2. I can strip it out and we can merge with just the JSON validation for now and I can open up an issue to add in a bit more scalable xml identification method in a subsequent PR

Thoughts @leehinman? I'm willing to go with either of the above, just let me know.

Reply:
so re-reviewing the net/http sniffing code -- it does look like the sniffer already handles cases starting with `<?xml` -- so if we did decide to remove this for the first pass, we'd still have some xml support, just not if the payload was missing the initial signature (which, as noted is the case in many xml-based api responses)

Reply:
Ok, how about we leave it in and open a PR to come up with a better XML validation?  If it is too much of a performance hit users can disable the processor.

I think the rules we need for XML are:

 - single root element that contains all other elements.
 - elements must be properly nested
 - elements must have closing tag or be “self-closing” (<tag/>)

We wouldn't know if it is "valid" XML since we don't have a schema for the XML, but anything that matches should probably be considered as XML.

Reply:
I like that idea -- I'll open up an issue and link it here after I address the other two real quick.

Reply:
Filed https://github.com/elastic/beats/issues/22965